_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,267 @@
+
+Before starting the migration, the `Kafka` custom resource must have the `strimzi.io/node-pools: enabled` annotation to use node pools, as well as the `strimzi.io/kraft: disabled` annotation that defines a ZooKeeper-based cluster.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: disabled # states that the cluster is ZooKeeper-based
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+Using the Unidirectional Topic Operator and a Kafka version greater or equal 3.7.0 are important prerequisites as well.
+
+Before migrating, a deployed cluster generally has severals pods running for the Kafka brokers and ZooKeeper nodes.
+
+```shell
+NAME                                          READY   STATUS    RESTARTS   AGE
+my-cluster-entity-operator-555785d699-kksrs   2/2     Running   0          17m
+my-cluster-kafka-0                            1/1     Running   0          18m
+my-cluster-kafka-1                            1/1     Running   0          18m
+my-cluster-kafka-2                            1/1     Running   0          18m
+my-cluster-zookeeper-0                        1/1     Running   0          20m
+my-cluster-zookeeper-1                        1/1     Running   0          20m
+my-cluster-zookeeper-2                        1/1     Running   0          20m
+strimzi-cluster-operator-7ddf57685d-245pd     1/1     Running   0          17m
+```
+
+#### Migrating from ZooKeeper to KRaft-based brokers
+
+The initial migration step involves deploying a `KafkaNodePool` custom resource to provision the KRaft controllers. 
+Then, you change the `strimzi.io/kraft` annotation from `disabled` to `migration` to initiate the process.
+
+Here's a snippet of the `KafkaNodePool` custom resource needed to deploy and activate the KRaft controllers:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  roles:
+    - controller
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: false
+```
+
+The corresponding `Kafka` custom resource is then updated to set the `strimzi.io/kraft` annotation to `migration`:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: migration # updated from "disabled" to "migration" to trigger the migration process
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+You can do that by annotating the custom resource.
+
+```shell
+kubectl annotate kafka my-cluster strimzi.io/kraft="migration" --overwrite
```

Comment:
I personally would change the order for the `kubectl annotate` command and the YAML example. But up to you.

Reply:
I like the current state, like showing how the resource should be and right after saying how to do that. I could change the order it's not a big deal but I should rephrase even the text around.

Reply:
Well, as I said, your call. But having the annotate command first would allow you to remove the ugly comment that doesn't fit the screen without scrolling as the annotate command would tell the people what to look for in the YAML.